InstantGrinder/InstantGrinder/InstantGrinderPlugin.cs

```diff
@@ -103,38 +192,120 @@ public void GridGridGroup(MyPlayer player, IEnumerable<MyCubeGrid> gridGroup)
                 .Field("block_count", blocks.Length)
                 .Write();
+
+            //MyHud.ChangedInventoryItems.Update();
+
+            //           if (ItemsMoved)
+            //               PlayerRefresh(player);
         }
 
-        void CopyItems(MyEntity src, MyInventory dst)
+        private void DeleteItems(MyEntity src)
         {
             if (!src.HasInventory) return;
 
+            string typeToDelete1 = "MyObjectBuilder_PhysicalGunObject";
+            string typeToDelete2 = "MyObjectBuilder_OxygenContainerObject";
+            string typeToDelete3 = "MyObjectBuilder_GasContainerObject";
+
             for (var index = 0; index < src.InventoryCount; index++)
             {
                 var inventory = src.GetInventory(index);
-                if (inventory.Empty()) continue;
+                if (inventory.Empty())
+                    continue;
+
+                // clean tools from source grid.
+                try
+                {
+                    var guns = src.GetInventory().ItemCount;
+                    for (var y = 0; y < guns; y++)
+                    {
+                        var ii = inventory.GetItemAt(y);
+                        if (ii.HasValue && ii.Value.Type.TypeId == typeToDelete1)
+                        {
+                            inventory.RemoveItemsAt(y, null);
+                            y--;
+                            guns--;
+                        }
+                    }
+
+                    var Oxygen = src.GetInventory().ItemCount;
+                    for (var y = 0; y < Oxygen; y++)
+                    {
+                        var ii = inventory.GetItemAt(y);
+                        if (ii.HasValue && ii.Value.Type.TypeId == typeToDelete2)
+                        {
+                            inventory.RemoveItemsAt(y, null);
+                            y--;
+                            Oxygen--;
+                        }
+                    }
+
+                    var Hydrogen = src.GetInventory().ItemCount;
+                    for (var y = 0; y < Hydrogen; y++)
+                    {
+                        var ii = inventory.GetItemAt(y);
+                        if (ii.HasValue && ii.Value.Type.TypeId == typeToDelete3)
+                        {
+                            inventory.RemoveItemsAt(y, null);
+                            y--;
+                            Hydrogen--;
+                        }
+                    }
+                }
+                catch
+                {
+                    return;
+                }
+            }
+        }
+
+        private void CopyItems(MyEntity src, MyInventory dst)
+        {
+            if (!src.HasInventory) return;
+ 
+            for (var index = 0; index < src.InventoryCount; index++)
+            {
+				var inventory = src.GetInventory(index);
+                if (inventory.Empty())
+                {
+                    GridInventoryIsEmpty = true;
```

Comment:
Maybe you can return this bool value from this method instead of a global state.